working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1863,52 +1863,108 @@ appears:
 
     The guard expression is evaluated in its case's case scope.
 
+    The body of a switch statement or expression case is executed in the case
+    body scope, defined below.
+
     The then statement of an if-case statement is executed in the case's case
     scope.
```

Comment:
I'd prefer to do it in the body scope too, and make the variables of the case scope always be final.
(It's not important for `if` and single-`case` switch cases, but I like the consistency of doing it the same way everywhere. Having 
```dart
switch (e) { 
  case int e when ++e > 0: print(e); 
  case int e when --e < 0: print(e); 
}
```
work one way and
```dart
switch (e) { 
  case int e when ++e > 0:
  case int e when --e < 0: print(e); 
}
```
work another is error-prone.
So, don't reuse the case scope, even when you can.

Reply:
Made it an error to assign to case variables in the guard, so those examples are prohibited directly. Even so, introducing new variables for the body is still visible because the guard can capture the variable (which is different from the body's variable). Since that's visible and pretty weird (but will almost never be seen), I'd prefer to only do that when we need to because there's a shared case.

I think I could reasonably explain to a user why:

```dart
        Function captured;

        bool capture(Function closure) {
          captured = closure;
          return true;
        }

        switch (['before']) {
          case [String a] when capture(() => print(a)):
          case [_, String a]:
            a = 'after';
            captured();
        }
```

Prints "before". The `a` in the body *can't* be the "same" `a` is in the guard because there's clearly *two* `a` variable declarations.

But I'd have a real hard time explaining or justifying why this would print "before":

```dart
        Function captured;

        bool capture(Function closure) {
          captured = closure;
          return true;
        }

        switch (['before']) {
          case [String a] when capture(() => print(a)):
            a = 'after';
            captured();
        }
```

There's absolutely no need for the `a` in the body to be different. So since we can, I think it's best to actually make it the same variable. There's only one variable declaration. There's nothing to merge.